### ISSUE TITLE
Syncthing-Fork v1.0.0.1 fails to open "Disable Battery Optimization" on Shield TV

### ISSUE THREAD

**Issue by o-l-a-v** [Catfriend1/syncthing-android#192]
If you have issues with this fork, feel free to post them. I'll look into them if a good reproducer or log is provided.

### Description of the issue
Syncthing-Fork v1.0.0.1 fails to open "Disable Battery Optimization" on nVidia Shield Android TV with Android 8.0.0, Experience Update v7.2.1.

Here's a [YouTube video](https://www.youtube.com/watch?v=4BMiW_REBfs) showing the very issue.

I've mentioned the issue for sniper at [nVidia GefForce forums](https://forums.geforce.com/default/board/159/), he says there's been no recent code changes related to that popup/ dialog in Shield TV firmware. Might never functioned properly ever since Shield got updated to Oreo, but I don't know that.

### Reproducer
* Install Syncthing-Fork on nVidia Shield Android TV running Shield Experience Update 7.2.1.
* Go through the initialization wizard. You won't get past the step regarding disabling Android battery optimization for Syncthing-Fork. Syncthing official android wrapper also fails here.

### Version Information
    App Version: 1.0.0.1
    Syncthing Version: 1.0.0.0
    Android Version: Android 8.0.0, Shield Experience Update 7.2.1

### Android Log
    adb shell / logcat
    or MatLog (search on GitHub) output
*MatLog won't run on Android TV*

**Comment by Catfriend1** [comment]
Thanks for the video and report. I'll Google if I can find a good solution to this. 
Did you try going to Android settings> battery/power and exclude the app there manually from power optimization? Did it work then?
Should I build a version of the app that is for testing and Skips the welcome slide for us to see how it works then?

**Comment by o-l-a-v** [comment]
There is no such menu to be found on Android TV (at least not on nVidia Shield), else there would be an easy solution to this problem.

Nah, no need to build a special version. My guess is bugs in nVidia firmware, but I'm not qualified to say / got no evidence.

There might be some ADB commands one could use, I'll research that.
Will also try to sideload MatLog.

**Comment by o-l-a-v** [comment]
Added a [thread](https://forums.geforce.com/default/topic/1092750) on nVidia forums, might help us to find out wether this is an Syncthing-Fork issue, or an issue with nVidia firmware.

**Comment by Catfriend1** [comment]
If possible, a log would be very interesting. I guess Nvidia has removed the power saving dialogs from android core. Dumping exact device info via adb (manufacturer and model especially) would help fix this issue in the future on our side.

**Comment by o-l-a-v** [comment]
I'm looking into getting logs out of it now. ADB over WiFi seems like a possibility, but every new app to test requires sideloading. A bit cumbersome, but I think I'll get it eventually.

**Comment by o-l-a-v** [comment]
Got some logs with MatLog, please tell if you can get any usefull information out of it.

**Comment by o-l-a-v** [comment]
Found a way to whitelist an app from doze / battery optimization, seems like it did the trick for Syncthing-Android and Syncthing-Fork

    adb shell dumpsys deviceidle whitelist +

**Comment by Catfriend1** [comment]
Great, i'll look into them as soon as I'm back on the computer :-).

**Comment by Catfriend1** [comment]
The work around should be in the tips & tricks collection and only displayed on Nvidia devices.

**Comment by Catfriend1** [comment]
Can you confirm this is correct:
adb shell dumpsys deviceidle whitelist +com.github.catfriend1.syncthingandroid
Which commands had to be entered before that to get adb over wifi access? Did you need to turn on developer options beforehand? I'm asking because I want to describe that a accurate as possible in tips &tricks section.

**Comment by o-l-a-v** [comment]
Had to enable developer menu, enable USB debugging then debugging over WiFi.
Got no info on ip:port, then connected pc to Shield USB Micro
Got ip:port info for ADB over WiFi, disconnected USB cable.
Adb server-kill, then adb server-start
Then adb connect ip:port
Got confirmation on Shield, confirmed/ gave adb permission to my PC.
Then i ran the command.

**Comment by o-l-a-v** [comment]
Tried again, it's simpler than what I first described:

* Enable Developer option
* Enable "USB debugging" and then "Network debugging" inside Developer options
* Wait for ip:port to show
* Locate / Download Google platform-tools, navigate to folder where adb.exe resides. Open Command Prompt here
* Type the following:
    adb start-server
    adb connect ip:port (example: adb connect 192.168.1.76:5555)
* Enable "Allow Network debugging" on the prompt that pops up on the nVidia Shield Android TV
* Then type the following to check weather you are successfully connected. Should state one device, and "connected"
    adb devices
* To whitelist Syncthing-Fork from doze / Android Battery Optimization, type the following
    adb shell dumpsys deviceidle whitelist +com.github.catfriend1.syncthingandroid
* If you ever want to revert this change, type the following in ADB
    adb shell dumpsys deviceidle whitelist -com.github.catfriend1.syncthingandroid

**Comment by Catfriend1** [comment]
Added wiki article: https://github.com/Catfriend1/syncthing-android/wiki/nVidia-Shield-TV-preparations
Thanks for writing this up so detailed 👍

**Comment by Catfriend1** [comment]
The problem lies here - as I guessed, NVIDIA shields is an Android OS "fork" which got those battery optimization menus patched out.

> ActivityManager(3560): START u0   {act=android.settings.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS   dat=package:com.github.catfriend1.syncthingandroid   cmp=com.android.tv.settings/.EmptyStubActivity}

```
01-04 18:45:18.890   I/am_pause_activity(3560):   [0,15218535,com.github.catfriend1.syncthingandroid/com.nutomic.syncthingandroid.activities.FirstStartActivity]
01-04 18:45:18.892   I/am_on_paused_called(9106):   [0,com.nutomic.syncthingandroid.activities.FirstStartActivity,handlePauseActivity]
01-04 18:45:18.895   I/am_restart_activity(3560):   [0,140226358,304,com.android.tv.settings/.EmptyStubActivity]
01-04 18:45:18.900   I/am_set_resumed_activity(3560):   [0,com.android.tv.settings/.EmptyStubActivity,minimalResumeActivityLocked]
01-04 18:45:18.900 D/ViewRootImpl(9106):   windowFocusChanged false com.github.catfriend1.syncthingandroid
01-04 18:45:18.906   I/am_finish_activity(3560):   [0,140226358,304,com.android.tv.settings/.EmptyStubActivity,app-request]
01-04 18:45:18.905 I/zygote64(4307):   Looking for service vendor.nvidia.hardware.power@1.0::IPower/default
01-04 18:45:18.906   I/am_pause_activity(3560):   [0,140226358,com.android.tv.settings/.EmptyStubActivity]
01-04 18:45:18.917 D/ViewRootImpl(9106):   windowFocusChanged true com.github.catfriend1.syncthingandroid
01-04 18:45:18.918   D/ControllerMapper_S(4363): same binder for   com.github.catfriend1.syncthingandroid skipping
01-04 18:45:18.917   I/am_set_resumed_activity(3560):   [0,com.github.catfriend1.syncthingandroid/com.nutomic.syncthingandroid.activities.FirstStartActivity,resumeTopActivityInnerLocked]
01-04 18:45:18.918   D/ControllerMapper_S(4363): notifyAppName   com.github.catfriend1.syncthingandroid old   :com.github.catfriend1.syncthingandroid
01-04 18:45:18.919   I/am_resume_activity(3560):   [0,15218535,303,com.github.catfriend1.syncthingandroid/com.nutomic.syncthingandroid.activities.FirstStartActivity]
01-04 18:45:18.924   I/am_on_resume_called(9106):   [0,com.nutomic.syncthingandroid.activities.FirstStartActivity,RESUME_ACTIVITY]
01-04 18:45:18.925 I/zygote64(4307):   Looking for service vendor.nvidia.hardware.power@1.0::IPower/default
01-04 18:45:18.926 I/zygote64(4307):   Starting thread pool.
01-04 18:45:18.932   I/am_destroy_activity(3560):   [0,140226358,304,com.android.tv.settings/.EmptyStubActivity,finish-idle]
01-04 18:45:18.931   E/powerHAL::common(3248): Unknown power hint: 0x8
01-04 18:45:18.934   I/wm_task_removed(3560): [304,removeTask]
01-04 18:45:19.696   I/ActivityManager(3560): START u0   {act=android.settings.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS   dat=package:com.github.catfriend1.syncthingandroid   cmp=com.android.tv.settings/.EmptyStubActivity} from uid 10127
01-04 18:45:19.697   E/powerHAL::common(3248): Unknown power hint: 0x8
01-04 18:45:19.697   I/am_create_activity(3560):   [0,170749710,305,com.android.tv.settings/.EmptyStubActivity,android.settings.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS,NULL,package:com.github.catfriend1.syncthingandroid,276824064]
```

**Comment by o-l-a-v** [comment]
Thanks, I'll request a fix from nVidia developers at their forum. 
It worked on previous Android TV version, Nougat.

**Comment by Catfriend1** [comment]
The fix from #192 will be released after merge.
